### DISCUSSION TITLE
WSL Win10/Win11 releases - confusing issue

### DISCUSSION THREAD

**Discussion by the-moog** [microsoft/WSL#9034] [General] [answered]
I'm going to come across as a confused noob, but I've searched and I can't find a definitive answer.

This is a request for somebody to 'fix' that confused noob.   Though I expect I am not the only one that has the same bug.

Our organisation (and it's a very large one) have no plans to move to Win10 any time soon (as that would be 10's of thousands of PCs) and we are experiencing increaseing isssues with what appears to be a very old WSL2 (e.g. some apps generate vmmem memory leaks which run into 30+Gb)

There are many Win11 WSL features we are all waiting for down in Win10 land.  A quick search yields many outstanding backport requests (WSLg, native EXT4, etc) and numerous bugs that are fixed in later versions.

afaik there is only one repo for WSL, yet there are at least three concurrent versions (that I know of)
 - WSL1
 - WSL2/Win10
 - WSL2/Win11

(I expect there are others), each dependent on the release platform for Windows (21H2, etc).  Combine that with multiple kernels and multiple OSs running on those kernels and we have a lot of combinations.

I can see this repo [https://github.com/microsoft/WSL](https://github.com/microsoft/WSL) only has one branch named 'master'

Could somebody explain the release versions of WSL (e.g. 0.70.0) (there are 20 prior releases) combined with what WSL2-Kernel releases work with what platform at what revision.

Perhaps the releases should indicate compatability and perhaps there should be more than one repo (or more than one branch) to express this.  At least a table in the wiki?.... [https://github.com/microsoft/WSL/wiki](https://github.com/microsoft/WSL/wiki)

**Comment by Biswa96** [comment]
The release version of WSL (e.g. 0.70.0) falls into WSL2/Win11 category. That version has no relation with Linux kernel version used within WSL2. The Linux kernel version follows stable releases.

**Comment by NotTheDr01ds** [comment]
Would recommend this be moved to [Discussions](https://github.com/microsoft/WSL/discussions) when a WSL team-member sees it.

> afaik there is only one repo for WSL
>
> I can see this repo https://github.com/microsoft/WSL only has one branch named 'master'

This repo *is* a bit different from most other Github repos since it does *not* actually contain any source code for WSL.  WSL is not an open-source project.  The *primary* use of this repo, as noted on the [main page](https://github.com/microsoft/WSL) is for:

* Reporting of Issues
* Discussions

> Our organisation .... have no plans to move to Win10 any time soon

I assume that's a typo and you mean *"move to Win 11"* (no worries, I probably have some typos in here, too).

> There are many Win11 WSL features we are all waiting for down in Win10 land. A quick search yields many outstanding backport requests (WSLg, native EXT4, etc) and numerous bugs that are fixed in later versions.

Microsoft has been very clear (although I probably can't find the post at the moment since it's been more than a year ago) that WSLg will not be backported to Win10.  To paraphrase my understanding and recollection of the post, there were architectural changes in Windows 11 that were required for WSLg.  It simply cannot run without those changes.

And I think it should be very clear by now that Microsoft will still be supporting Windows 10 going forward, but most likely only with fixes, especially security.  If there's a new feature going into Windows 10 at this point (and I don't know that there has been), it's either strategic (as a hypothetical for-instance, some Azure integration feature) or a contractual commitment for features that have not yet been delivered (deferred revenue rules).  Again, this is just my conjecture based on previous experience in this space.

> Could somebody explain the release versions of WSL (e.g. 0.70.0) (there are 20 prior releases) combined with what WSL2-Kernel releases work with what platform at what revision.
>
> Perhaps the releases should indicate compatability and perhaps there should be more than one repo (or more than one branch) to express this. At least a table in the wiki?

Not really a need for a table.  It's been a steady-progression for WSL releases:

* WSL1:  The last release appears to be the one that was delivered in-box with Windows 11 22000 (the initial release last year).  It is unknown whether Microsoft plans any future development here.  As far as I can tell, there have been no WSL1-specific issues here that have been fixed in the last 12 months.  Even some critical failures (Ubuntu 22.04 was pretty much completel broken on WSL1 out-of-the-box for 6 months) received no Microsoft response.  Based on this, I'm personally assuming that WSL1 is EoL unless/until I see someone from Microsoft say otherwise (or confirm).

    This isn't a complaint, really -- Software projects sometimes come to an end, especially when/if there's a suitable replacement (good riddance to Console Host!  Long live Terminal! 😀).  It would just be nice to know.  IHMO, if Microsoft can improve WSL2's 9P performance (used to access `/mnt/c`, etc.) then I can personally work-around any other WSL2 "issues" like networking.

* WSL2:  There are two versioning schemes for WSL2 releases:

    1. Through Windows 10's lifecycle, new features were added to WSL (and thus WSL2) only in conjunction with a corresponding Windows release.  These could either be an Insider release, or, for most of us, a H1/H2 cadence-Windows release.  For this reason, WSL releases (both WSL1 and WSL2) were typically known by their corresponding Windows release (e.g. 10.0.19042 that came with Windows 10 20H2, if Wikipedia has the [release number correct](https://en.wikipedia.org/wiki/Windows_10_version_20H2)).

        Those are now called (or perhaps were always internally called) "inbox" releases.  It's a new term to me, at least, as it just recently showed up in the latest `wsl --install` [helptext](https://github.com/microsoft/WSL/discussions/8975).

    2. The last GA (general availability) inbox release was in Windows 11 22000 (the initial GA).  At that point, the team was (mostly) able to decouple the delivery/installation of WSL from Windows itself.  In October (I believe) of last year, we began seeing "Windows Subsystem for Linux Preview" releases in the Microsoft Store.  Again, these are only for Windows 11 users, since the functionality (including and especially WSLg) is tied to Windows 11.

        These are the releases you see in the [Releases](https://github.com/microsoft/WSL/releases) page here.  My understanding is that Windows Insider releases automatically push the latest "Preview".  However, [Windows 11 22H2 (22621) comes with the previous inbox](https://github.com/microsoft/WSL/discussions/8936) that was in 22000/21H2.

        These are really WSL2-only releases.  To add WSL1 support, you still need to turn it on via the "Turn Windows features on or off" (or CLI equivalent).  `wsl --install` used to do this automatically, but that has changed in 22621/22H2 so that you now specifically have to `wsl --install --enable-wsl1.

Kernel releases, as @Biswa96 mentioned, are somewhat decoupled from this.  You should be able to use any recent WSL2 kernel even under Windows 10.  This allows, for instance, USB/IP support on Windows 10.  So kernel-only features (which are rare) *can* be enabled on Windows 10.

> There are many Win11 WSL features we are all waiting for down in Win10 land. A quick search yields many outstanding backport requests (WSLg, **native EXT4**, etc)

Addressing specifically the ext4 comment here -- WSL2 has *always* had native ext4 support, even from the first release on Windows 10.  So not quite sure what you mean about that feature.

**Reply by the-moog** [reply]
Thanks for the detailed answer and it does go some way to explaining the situation.  And yes I did mean upgrade to Win11, oops.  I do remember some time back seeing a post some place saying that there are technical reasons why WSLg will never appear in WIN10 (sadly), so I was also aware of that.

My question really relates to bug fixes.
There is a bug unrelated to WSLg (though probably fixed due to it being a thing) that appears to be fixed in 0.70.0 that is affecting work I am doing.  Specifically some apps eat memory very quickly and the vmms service is the culprit.   It seems to be related to graphical apps in XWindows as I have not seen this with the non-gui version of the app I am using. (Xilinx Vivado if you need to know) The memory usage goes to 16G eventually resulting in a huge system crash.  I tried installing 0.70.0 and found it's only for Win11.  So there is some prior or future release for Win10 that may fix this issue when the fix for Win11 is backported.  Hence my question.  IMO There should be a clear version vs platform but that seems not to be the case?

**Comment by NotTheDr01ds** [comment]
@the-moog I do agree, I didn't quite answer your question, even with all that ;-).  So let me try again.

First, let me start with the bad news:

> So there is some prior or future release for Win10 that may fix this issue when the fix for Win11 is backported

It's extremely unlikely, in my opinion, that you'll see the fix backported to a WSL release on Windows 10.  I do not and cannot speak for anyone at Microsoft, but they typically don't port fixes or features backwards (for any product) unless there's a strategic (e.g. perhaps something in Windows broke specifically on Azure), contractual, or security reason to do so.

The last time we saw any mention of a new feature or fix on WSL on Windows 10 was 21H2.  As far as I can tell, the WSL version in Windows 10 22H2 is exactly the same as it was in 21H2.  Further, the WSL "in box" version of WSL in Windows 11 22H2 is exactly the same as it was in Windows 11 21H2.

> There should be a clear version vs platform but that seems not to be the case?

I agree it's confusing.  There are several reasons for this:

* In the "old" versioning scheme for WSL, the WSL release was tied to a Windows build number.  These builds were often Insider/Preview/Fast-ring/etc. builds, and untangling those version numbers requires a secret decoder ring that I just don't have.

* At some point, development shifted from Windows 10 to Windows 11.  There are a few features mentioned in the [release notes](https://learn.microsoft.com/en-us/windows/wsl/release-notes) that *were* developed against both Windows 10 and Windows 11, since it was a transitionary period.  I'm thinking specifically of things like GPU support, which was made available on both Windows 10 21H2 and Windows 11 21H2.  Other features, like `wsl --mount` and WSLg, however, were only released for Windows 11 21H2.

* And then, after Windows 11 was released, the versioning scheme (and entire release mechanism) changed.  Up to 21H2, WSL releases were tied to a Windows release ("in box").  Immediately after 21H2, WSL was decoupled from Windows and released via the Store (or direct download from the releases page here).  At that point, the [versioning scheme (and release notes) shifted to the model used today](https://github.com/microsoft/WSL/releases).

    Again, it's important to understand that *all* of [these "modern"/Store releases](https://github.com/microsoft/WSL/releases) are for Windows 11 only, and they are WSL2-only (no WSL1 support unless it's installed separately).  They'll all work on Windows 11 21H2 or 22H2, but none of the releases on that page will work on Windows 10, and I have no expectation that they ever will.

As for the kernel, the latest that is available for direct *update* on Windows 10 appears to me to be 5.10.102.1.  However, there's no reason that I'm aware of that you can't use any later release of the kernel even on Windows 10.  I'll test later myself, but you should be able to simply copy over the kernel from a Windows 11 system to a Windows 10 system and create/update your `%userprofile%\.wslconfig` to point to it via the `kernel` setting per the [configuration guide](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#configuration-setting-for-wslconfig).

If the fix you need was in the kernel, then that may solve the problem.  Otherwise:

> Our organisation (and it's a very large one) have no plans to move to Win10 any time soon

I've worked for some pretty big ones myself (3 Fortune 500 companies), and in my experience, there's almost *always* a trial going with the next release for a subset of users.  Work with IT on the problem you are having, let them know that there is a critical fix in Windows 11 for an issue you are having, and ask if you can become a member of the beta/preview group.

If not (and it can be tough to get through the red-tape in IT, I know), play squeaky-wheel with your management until the issue gets pushed through.  If it's critical, then eventually something will break the log-jam, ... hopefully.

**Comment by NotTheDr01ds** [comment, marked as answer]
@the-moog Well, I'm happy to be wrong about some of the above.  WSL 1.0.0 is now available for Windows 10 with [KB5020030](https://support.microsoft.com/en-gb/topic/november-15-2022-kb5020030-os-builds-19042-2311-19043-2311-19044-2311-and-19045-2311-preview-237a9048-f853-4e29-a3a2-62efdbea95e2).  It's not quite clear yet whether all Windows 10 releases are supported or just 22H2, though, since KB5020030 applies to all (I believe) currently supported Windows 10 releases.

**Reply by the-moog** [reply]
Thanks, once again, for the above detailed clarification and the heads up on the new release. I am really forward to trying it out as I am sure others will be just as excited.

**Comment by the-moog** [comment]
I found it was not simple to update to Win10/WSLg as this issue https://github.com/microsoft/WSL/issues/9157 raised it's head. But after two windows update, first to install 22H2 and a second to install KB5020030  it seems to be working. I have to congratulate MS in rolling this out and thank the hard work from the devs that make this happen.  WSL has revolutionised the way we work.  I wonder if anybody here remember the previous aborted attempt at this some years back, I think it was called SUA (or something like that) and based around Cygwin and a complete dogs dinner - shame we did not have WSL back then.....
